File: tests/unit/CorrectionTests.cpp
#include <QtTest>
#include <QSignalSpy>
#include "domain/Correction.h"
#include "domain/ChecklistItem.h"

class CorrectionTests : public QObject
{
    Q_OBJECT

private slots:
    void testDefaults();
    void testNameProperty();
    void testDateProperty();
    void testAddItem();
    void testRemoveItem();
    void testGetItem();
    void testItemCount();
    void testAddNewItem();
    void testSignals();
};

void CorrectionTests::testDefaults()
{
    Correction corr;
    QVERIFY(!corr.id().isEmpty());
    QCOMPARE(corr.name(), QString("Sin nombre"));
    QVERIFY(!corr.date().isEmpty());
    QCOMPARE(corr.itemCount(), 0);
}

void CorrectionTests::testNameProperty()
{
    Correction corr("Mi Corrección");
    QCOMPARE(corr.name(), QString("Mi Corrección"));

    corr.setName("Actualizada");
    QCOMPARE(corr.name(), QString("Actualizada"));
}

void CorrectionTests::testDateProperty()
{
    QDate today(2024, 1, 15);
    Correction corr("Test", today);
    QCOMPARE(corr.date(), QString("2024-01-15"));
}

void CorrectionTests::testAddItem()
{
    Correction corr;
    ChecklistItem *item = new ChecklistItem("Item 1");
    corr.addItem(item);

    QCOMPARE(corr.itemCount(), 1);
    QCOMPARE(corr.getItem(0)->text(), QString("Item 1"));
}

void CorrectionTests::testRemoveItem()
{
    Correction corr;
    corr.addNewItem("Item 1");
    corr.addNewItem("Item 2");
    QCOMPARE(corr.itemCount(), 2);

    corr.removeItem(0);
    QCOMPARE(corr.itemCount(), 1);
    QCOMPARE(corr.getItem(0)->text(), QString("Item 2"));
}

void CorrectionTests::testGetItem()
{
    Correction corr;
    corr.addNewItem("Item 1");

    ChecklistItem *item = corr.getItem(0);
    QVERIFY(item != nullptr);
    QCOMPARE(item->text(), QString("Item 1"));

    // Out of bounds
    QVERIFY(corr.getItem(-1) == nullptr);
    QVERIFY(corr.getItem(1) == nullptr);
}

void CorrectionTests::testItemCount()
{
    Correction corr;
    QCOMPARE(corr.itemCount(), 0);

    corr.addNewItem("A");
    QCOMPARE(corr.itemCount(), 1);

    corr.addNewItem("B");
    QCOMPARE(corr.itemCount(), 2);
}

void CorrectionTests::testAddNewItem()
{
    Correction corr;
    corr.addNewItem("New item");

    QCOMPARE(corr.itemCount(), 1);
    QCOMPARE(corr.getItem(0)->text(), QString("New item"));
    QCOMPARE(corr.getItem(0)->done(), false);
}

void CorrectionTests::testSignals()
{
    Correction corr;
    QSignalSpy nameSpy(&corr, &Correction::nameChanged);
    QSignalSpy countSpy(&corr, &Correction::itemCountChanged);

    corr.setName("New name");
    QCOMPARE(nameSpy.count(), 1);

    corr.addNewItem("Item");
    QCOMPARE(countSpy.count(), 1);
}

QTEST_MAIN(CorrectionTests)
#include "CorrectionTests.moc"

File: csharp/tests/Checklist.Core.Tests/CorrectionTests.cs
using Checklist.Domain;
using Xunit;

namespace Checklist.Core.Tests;

public class CorrectionTests
{
    [Fact]
    public void Defaults()
    {
        var corr = new Correction();
        Assert.False(string.IsNullOrEmpty(corr.Id));
        Assert.Equal("Unnamed", corr.Name);
        Assert.False(string.IsNullOrEmpty(corr.Date.ToString("yyyy-MM-dd")));
        Assert.Equal(0, corr.ItemCount);
    }

    [Fact]
    public void NameProperty()
    {
        var corr = new Correction("Mi Corrección");
        Assert.Equal("Mi Corrección", corr.Name);

        corr.Name = "Actualizada";
        Assert.Equal("Actualizada", corr.Name);
    }

    [Fact]
    public void DateProperty()
    {
        var today = new DateOnly(2024, 1, 15);
        var corr = new Correction("Test", today);
        Assert.Equal("2024-01-15", corr.Date.ToString("yyyy-MM-dd"));
    }

    [Fact]
    public void AddItem()
    {
        var corr = new Correction();
        var item = new ChecklistItem("Item 1");
        corr.AddItem(item);

        Assert.Equal(1, corr.ItemCount);
        Assert.Equal("Item 1", corr.GetItem(0)?.Text);
    }

    [Fact]
    public void RemoveItem()
    {
        var corr = new Correction();
        corr.AddNewItem("Item 1");
        corr.AddNewItem("Item 2");
        Assert.Equal(2, corr.ItemCount);

        corr.RemoveItem(0);
        Assert.Equal(1, corr.ItemCount);
        Assert.Equal("Item 2", corr.GetItem(0)?.Text);
    }

    [Fact]
    public void GetItem()
    {
        var corr = new Correction();
        corr.AddNewItem("Item 1");

        var item = corr.GetItem(0);
        Assert.NotNull(item);
        Assert.Equal("Item 1", item!.Text);

        Assert.Null(corr.GetItem(-1));
        Assert.Null(corr.GetItem(1));
    }

    [Fact]
    public void ItemCount()
    {
        var corr = new Correction();
        Assert.Equal(0, corr.ItemCount);

        corr.AddNewItem("A");
        Assert.Equal(1, corr.ItemCount);

        corr.AddNewItem("B");
        Assert.Equal(2, corr.ItemCount);
    }

    [Fact]
    public void AddNewItem()
    {
        var corr = new Correction();
        corr.AddNewItem("New item");

        Assert.Equal(1, corr.ItemCount);
        Assert.Equal("New item", corr.GetItem(0)?.Text);
        Assert.False(corr.GetItem(0)?.Done);
    }

    [Fact]
    public void Signals()
    {
        var corr = new Correction();
        var nameCount = 0;
        var countCount = 0;
        corr.NameChanged += () => nameCount++;
        corr.ItemCountChanged += () => countCount++;

        corr.Name = "New name";
        Assert.Equal(1, nameCount);

        corr.AddNewItem("Item");
        Assert.Equal(1, countCount);
    }
}